webapp/channels/src/components/tooltip/index.tsx

```diff
@@ -0,0 +1,200 @@
+    }),
+    cmd: '⌘',
+    ctrl: defineMessage({
+        id: 'shortcuts.generic.ctrl',
+        defaultMessage: 'Ctrl',
+    }),
+    option: '⌥',
+    shift: defineMessage({
+        id: 'shortcuts.generic.shift',
+        defaultMessage: 'Shift',
+    }),
+};
+
+interface Options {
+
+    /**
+     * The placement of the tooltip. Defaults placement is top.
+     */
+    placement?: Placement;
+
+    /**
+     * Whether to automatically determine the placement of the tooltip.
+     * ref: https://floating-ui.com/docs/autoPlacement
+     */
+    autoPlacement?: boolean;
+
+    /**
+     * The allowed placements of the tooltip. This should be used in conjunction with `autoPlacement`.
+     */
+    allowedPlacements?: Placement[];
+
+    /**
+     * Callback fired when the tooltip is opened or closed.
+     */
+    onChange?: (open: boolean) => void;
+}
+
+interface Props {
+    children: ReactNode;
+    options?: Options;
+    title: string | ReactNode;
+    emoticon?: string;
+    isEmoticonLarge?: boolean;
+    hint?: string;
+    shortcut?: ShortcutDefinition;
+}
+
+function Tooltip(props: Props) {
+    const [open, setOpen] = useState(false);
+
+    const arrowRef = useRef(null);
+
+    function handleChange(open: boolean) {
+        setOpen(open);
+
+        if (props.options?.onChange) {
+            props.options.onChange(open);
+        }
+    }
+
+    const {refs: {setReference, setFloating}, floatingStyles, context} = useFloating({
+        open,
+        onOpenChange: handleChange,
+        placement: props.options?.placement ?? DEFAULT_PLACEMENT,
+        whileElementsMounted: autoUpdate,
+        middleware: [
+            props.options?.autoPlacement ? autoPlacement({
+                allowedPlacements: props.options?.allowedPlacements ?? DEFAULT_ALLOWED_PLACEMENTS,
+            }) : null,
+            offset(ARROW_OFFSET),
+            arrow({
+                element: arrowRef,
+            }),
+        ],
+    });
+
+    const hover = useHover(context, {
+        restMs: TOOLTIP_REST_TIME_BEFORE_OPEN,
+        delay: {
+            open: Constants.OVERLAY_TIME_DELAY,
+        },
+    });
+    const focus = useFocus(context);
+    const dismiss = useDismiss(context);
+    const role = useRole(context, {role: 'tooltip'});
+
+    const {getReferenceProps, getFloatingProps} = useInteractions([hover, focus, dismiss, role]);
+    const {isMounted, styles: transitionStyles} = useTransitionStyles(context, {
+        duration: {
+            open: TOOLTIP_APPEAR_DURATION,
+            close: TOOLTIP_DISAPPEAR_DURATION,
+        },
+        initial: {
+            opacity: 0,
+        },
+        common: {
+            opacity: 1,
+        },
+    });
+
+    const tooltipTrigger = useMemo(() => {
+        if (!isValidElement(props.children)) {
+            // eslint-disable-next-line no-console
+            console.warn('Tooltip must have a valid child element');
+            return null;
+        }
+
+        return cloneElement(
```

Comment:
Because we're cloning the child here, we're going to need to make it so that anything passed as the child of a `Tooltip` will accept arbitrary props because `getReferenceProps` will be injecting various event handlers and aria props from `useHover`, `useFocus`, `useDismiss`, and `useRole`. `getReferenceProps` handles merging those with the props that are already passed into the child, but the child needs to accept those to pass them on to whatever it renders. That's fine if whatever uses this component just renders HTML, but if it renders to another child component, that might cause problems similar to the ones that we've run into with the MUI Menu.
```jsx
// This works because the extra props are passed directly into the span
function Works1() {
    return (
        <Tooltip>
            <span>{'tooltip'}</span>
        </Tooltip>
    );
}

// This works because extra props are explicitly handled by the custom child component
function Works2() {
    function Works2Child({text, ...otherProps}: ChildProps) {
        return <span {...otherProps}>{text}</span>;
    }

    return (
        <Tooltip>
            <Works2Child text='tooltip'/>
        </Tooltip>
    );
}


// This doesn't work because extra props aren't passed on by the custom child component
function DoesntWork() {
    function DoesntWorkChild({text}: ChildProps) {
        return <span>{text}</span>;
    }

    return (
        <Tooltip>
            <DoesntWorkChild text='tooltip'/>
        </Tooltip>
    );
}
```

I don't think we've run into that issue with `OverlayTrigger` since I'm pretty sure it add its own wrapper `div` around the reference element which it can pass those event handlers to, but I feel like I've seen that issue with some of the other React Bootstrap elements as well.

That's not a massive problem, but it does mean that anyone using this component needs to be aware of the extra potential risk.

I'm trying to think if another pattern like using a render prop (or passing a function as a child) would be better here since it would make it explicit that `Tooltip` adds extra props to whatever component is being used with it. I did find that custom hooks are apparently preferable to render props, but then we'd lose a lot of the niceness of just rendering `<Tooltip></Tooltip>` has.

https://react.dev/reference/react/cloneElement#alternatives

Any thoughts on that?

CC @larkox since I know you're big on using hooks

Reply:
Also, given you're going to be out of office for a while, we can always change this approach later after you get back

Reply:
I discussed this with @M-ZubairAhmed and I am not a big fan of this approach. Among other things, we are using a memo that will be rendered useless due to `children` being a new object on every render.

The problem is that there are not many options that don't break the current interface.

If I remember correctly, the plan is to get the reference by just wrapping the content in a `span`  tag. That should not mess with styling (at least not much in many places) and still get us a reference with the correct location.

Reply:
Yes i wrapped it in span so far so UI changes. Will have to see when we adopt to other places as well

Reply:
Wrapping in a span won't work anywhere that we pass a div or other element that can't be nested inside of a span, and it could also cause styling issues anywhere that we rely on descendent selectors.

Looking at OverlayTrigger, it also clones the inner element ([link](https://github.com/mattermost/react-bootstrap/blob/0.34/src/OverlayTrigger.js#L317)), so that's why it works here. It also has the same issue though where the child element has to pass on the inner props as well though. Funnily enough, I just noticed that even our ancient version of React Bootstrap passes along almost the same props as Floating UI adds.

Let's go back to using `cloneElement` for now and continue this discussion later to avoid accidentally breaking anything. We can live with the downsides of `cloneElement` since they're no worse than we have with the current code, so let's not rush to fix that in a way that'll introduce other issues

Reply:
oh i see, @larkox should i wrap cloneElement in useMemo, if yes then would the children as dependencies would be not usefull since they always change

Reply:
Let's go with simple clone, no use memo or anything, and let's see what happens. I am not a big fan of the idea, but as @hmhealey it is not worse than what we currently have.